jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+**For chas**: **IMPORTANT:** there is a structural issue here where the period-determination code processes one big block of frame buffers all together, once. That is not a good solution when running live - it will be much better to make a new attempt to determine a period after each individual new frame is received, rather than doing a massive chunk of processing in the loop in establish_indices. I really think this needs a refactor. Otherwise, apart from anything else, we will end up acquiring significantly more frames than we need to before we can establish a period. There are also scenarios where this approach will fail, where it could succeed if it was allowed just to keep trying indefinitely, processing a rolling buffer of recent frames. (Minor related observation: once that is changed, and the loop is removed from this function, the "didn't find a period" error will not be "critical" when running live)
+
+prospective_optical_gating.py
+
+[No comments beyond the inline ones in the source code]
+
+emulate.py
+
+I haven't really read through this in detail yet, but I do remember that I didn't find it particularly helpful as an example to follow when I was trying to do my own offline analysis using this codebase. I need to go back and remind myself what the difficulties were that I encountered...
+
+stage_control_functions.py
+
+Looks pretty well commented and structured, but I would implement this as a class, so that other (sub-)classes could be substituted to control other stage types, etc. I will give this some more thought. **For me**: how is this used? Which object is the most sensible one to be in charge of all this?
+
+app
+
+For my benefit: base_camera.py runs a background thread which posts events whenever a frame is available. frames() uses "yield" to pass frame buffers to an iterator.
+
+app.gen() is a "video streaming generator function" which in turn "yield"s frames (in what looks like a MIME-like format) to the caller. I am not clear how video_feed behaves, though.
+
+Basic questions **for chas**:
+
+- What is the overall function of app.py? Do I understand correctly that these decorated functions are called in response to messages from the client webpage? In response to those messages, they appear to both perform internal state-affecting actions (i.e. run the synchronization algorithms, in some cases), *and also* then return the actual HTTP source(*) that should now be displayed in the webpage. Am I understanding that correctly? What is the deal with the "session" object? Could really do with a comment at the top of the module to explain what's going on with all of this.                (*) Oh, or sometimes raw video images apparently, in video_feed()...
```

Comment:
It sounds like you understand whats happening... I'll be honest I didn't comment anything that is just 'how flask works' but probably should have. The session object is basically client-side data, like a cookie.

Reply:
Yes - certainly for somebody like me, trying to make sense of it from scratch, I think it's worth a brief comment to explain the basic concept of what's going on, even if I'm sure all of the flask architecture is well documented somewhere. It's nice when I *don't* have to read a flask tutorial to be able to have a general picture of how the code built on top of it is going to be behave.